### ISSUE TITLE
android_prebuilt_aar should actually export its deps like prebuilt_jar

### ISSUE THREAD

**Issue by edisonw** [facebook/buck#328]
Now that Gradle supports transitive aars, there are libraries out there that requires a prebuilt aar to be depended on another prebuilt aar, there is no way to pass that in right now. 

Can we add support for the following pattern? 

android_prebuilt_aar {
      deps = [
          other prebuilt aars. 
      ]
}

**Comment by sdwilsh** [comment]
deps should work today.  What are you seeing when doing this?

**Comment by edisonw** [comment]
it was complaining about R resources not found.

**Comment by edisonw** [comment]
I'm using a buck from few weeks ago, let me update to latest and give it another try.

**Comment by edisonw** [comment]
Yea, with master I'm still getting BUILD FAILED: //:depdendant#aar_android_resource failed with exit code 1:
generate_resource_ids 


Setup: 

android_prebuilt_aar(
    name = 'depdendant',
    aar = 'depdendant/depdendant.aar',
    deps = [
      ':depdendant-core'
    ]
)

**Comment by sdwilsh** [comment]
I'm guessing you need `:dependant-core` also in whatever libraries need `:dependant`, yeah?  If so, we'll need to add `exported_deps` to the rule and you'd want to use that.

**Comment by edisonw** [comment]
^yea, correct, there are other libraries that need :dependent-core as well.

**Comment by sdwilsh** [comment]
Can you apply https://gist.github.com/sdwilsh/3d4645edc5d4f3bbf0a1 locally and see if it fixes it for you?  Writing a test is going to take a while for this, but if this unblocks you, I can just get it in.

**Comment by edisonw** [comment]
Umm...it won't get through the process stage now:

BUILD FAILED: //:dependent: exported dep //:dependent#aar_android_resource (android_resource) must be a type of java library.

**Comment by sdwilsh** [comment]
Alright, the naive fix won't do it.  Hopefully I can get this resolved next week.  For now, just include both in the deps of dependent libraries (which sucks, I know).

**Comment by sdwilsh** [comment]
Also, do you happen to have an example AAR that does this?

**Comment by rallat** [comment]
This [aar](https://twittersdk.artifactoryonline.com/twittersdk/public/com/digits/sdk/android/digits/1.5.0/) has transitive dependencies on another aars

**Comment by sdwilsh** [comment]
@rallat - looking over the `pom`, that seems to depend on twitter-core, but I'm not seeing that available for download on that site.  Any pointers?

@edisonw - if you want to give https://gist.github.com/sdwilsh/f98020f4573e506705b6 a try, I think it may do the trick in a way that doesn't fail like the above (I missed something with that quick diff last week).  The thing that I don't think will work, although I'm not sure if it matters, is that android resources won't be exported, so you likely wouldn't be able to use them in your libraries without declaring a first-order dependency.  I haven't used AARs much, so I'm wondering what you would expect the expected behavior to be in that situation.

**Comment by edisonw** [comment]
@sdwilsh thanks! I applied the diff from latest master (e96bb81d675d2517aee5e5fcc36c4833ca85a346) This did make the aar pass it's own build stage, but now prebuilt_jar are not found in android_lib projects (as if the prebuilt jars were not included). Any ideas? 

As for twitter-core, you can find it here: 

http://twittersdk.artifactoryonline.com/twittersdk/repo/com/twitter/sdk/android/twitter-core/1.3.4/

**Comment by sdwilsh** [comment]
Ahh, so you have a `android_prebuilt_aar` that has `prebuilt_jars` in its dependency chain?  That requires a bit more code than what I have, but I can make that work too.

Do you have some sample code that would use the right bits to demonstrate the bug?  I'm worried that I'm not going to write an integration test for this that will make sure we don't regress this behavior for you.

**Comment by edisonw** [comment]
"a android_prebuilt_aar that has prebuilt_jars in its dependency chain"
^Yea, tho the diff broke other android_libs that have prebuilt_jars in them for some reason. :| 

I'll see if I can use digits as a clean example. We do have pretty complicated setup right now with other jars as well.

**Comment by rallat** [comment]
here is twitter core: https://twittersdk.artifactoryonline.com/twittersdk/public/com/twitter/sdk/android/twitter-core/
the issue is that we are using R from the transitive dependency and buck isn't finding them

**Comment by sdwilsh** [comment]
I believe, without tests still, that https://gist.github.com/sdwilsh/b0e7ca14bda617041c11 should solve this.  Basically, the `android_prebuilt_aar` needs to export and `classes.jar` and any `prebuilt_jar`s that exist in its deps, which is what that change should do.

I'll spend some time today cleaning that up and getting it merged in.

**Comment by sdwilsh** [comment]
That's probably not going to work 100% either, but I just talked with @natthu and have a good idea on what the right fix is.

**Comment by edisonw** [comment]
Sounds good. Thanks!

**Comment by sdwilsh** [comment]
I filed #332 to track the `prebuilt_jar` case, which is actually a lot easier to deal with.

**Comment by sdwilsh** [comment]
What we have in the tree now might be enough for you, but it's not 100% fixed yet.

**Comment by sdwilsh** [comment]
#391 is going to close out the remaining piece of this

**Comment by edisonw** [comment]
Thanks!

**Comment by rallat** [comment]
Thanks!

**Comment by sdwilsh** [comment]
While #391 isn't closed, it's just adding tests now.  @natthu fixed the remaining part of this in 4cf0eb7a744c9583cc3f45f7952ef3aa6ae8afae.
